lib/howl/interactions/external_command.moon

```diff
@@ -0,0 +1,207 @@
+-- Copyright 2012-2015 The Howl Developers
+-- License: MIT (see LICENSE.md at the top-level directory of the distribution)
+
+glib = require 'ljglibs.glib'
+import app, interact, sys from howl
+import File from howl.io
+import Matcher from howl.util
+import ListWidget, markup from howl.ui
+import file_matcher, parse_path from howl.interactions.file_selection
+append = table.insert
+
+available_commands = ->
+  commands = {}
+  for path in sys.env.PATH\gmatch '[^:]+'
+    dir = File path
+    if dir.exists
+      append commands, child.basename for child in *dir.children
+
+  return commands
+
+text_parts = (text) ->
+  parts = [p for p in text\gmatch '%S+']
+  append parts, '' if text\ends_with ' '
+  parts
+
+last_text_part = (text) ->
+  parts = text_parts text
+  parts[#parts]
+
+get_cwd = ->
+  buffer = app.editor and app.editor.buffer
+  directory = buffer and (buffer.file and buffer.file.parent or buffer.directory)
+  directory or File glib.get_current_dir!
+
+looks_like_path = (text) ->
+  return unless text
+  for p in *{
+    '^%s*./',
+    '^%s*../',
+    '^%s*/',
+    '^%s*~/',
+  }
+    if text\umatch p
+      return true
+
+class ExternalCommandEntry
+  run: (@finish, @opts={}) =>
+    @command_line = app.window.command_line
+    @commands = nil
+    @list_widget = nil
+    @auto_show_list = true
+    local directory
+    if @command_line.directory and @opts.path
+      directory = @command_line.directory / @opts.path
+    elseif @command_line.directory
+      directory = @command_line.directory
+    elseif @opts.path
+      directory = File @opts.path
+    if not (directory and directory.exists and directory.is_directory)
+      directory = get_cwd!
```

Comment:
If a directory or path was actually specified, I think it makes sense to raise an error if the resulting directory does not exist, rather than silently ignoring it? Also styling, the `if not` stanza is not part of the preceding if/else chain so a blank line would improve readability.